hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/protocol/commands/ReconstructECContainersCommand.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.apache.hadoop.ozone.protocol.commands;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import com.google.protobuf.ByteString;
+import org.apache.hadoop.hdds.client.ECReplicationConfig;
+import org.apache.hadoop.hdds.protocol.DatanodeDetails;
+import org.apache.hadoop.hdds.protocol.proto
+    .StorageContainerDatanodeProtocolProtos.ReconstructECContainersCommandProto;
+import org.apache.hadoop.hdds.protocol.proto
+    .StorageContainerDatanodeProtocolProtos.ReconstructECContainersCommandProto
+    .Builder;
+import org.apache.hadoop.hdds.protocol.proto
+    .StorageContainerDatanodeProtocolProtos.SCMCommandProto.Type;
+
+import com.google.common.base.Preconditions;
+
+/**
+ * SCM command to request reconstruction of EC containers.
+ */
+public class ReconstructECContainersCommand
+    extends SCMCommand<ReconstructECContainersCommandProto> {
+
+  private final long containerID;
+  private final List<DatanodeDetails> sourceDatanodes;
+  private final List<DatanodeDetails> targetDatanodes;
+  private final byte[] missingContainerIndexes;
```

Comment:
Hi Uma, the info looks good, I wonder whether there are additional info needed, for example:
- ReplicaIndexes for sourceDatanodes, maybe we could learn this from the HealthyDNs in the group, but the `readContainer` grpc call don't provides this index for now, so if SCM carries this, it then we don't have to worry about it.
- The pipelineID may be needed to reconstruct the container, see `ContainerData#originPipelineId`. SCM manages the pipelineID, so it is reasonable for it to deliver this.

Reply:
Hi Uma, I'd like to add one more nit: `missingContainerIndexes` may better be just `targetIndexes` because when we recover each stripe we also have the concept of `missingIndexes`. But `missingContainerIndexes` and `missingIndexes on the stripe level` are different for partial stripes, so may be `targetIndexes` is better which just means those indexes are assigned to the `targetDatanodes`.

Reply:
@guihecheng Thank you for the review.
I added the replica indexes now. For the pipelineID, I am not sure we have any significance for the closed containers.

Coming to missing indexes:
Target nodes needed not be fixed with any indexes. As we try node per rack, there is no strict requirement to place the node with given index only.
But partial stripe case is very good point. Seems like we don't have a way for SCM to determine at stripe level. This is simply at the container level. Missing index means, the container which is missed at that index position. Stripe level determination should be done by recovery logic, whether that index is required to recover for that stripe or not based on block group len. 

I was thinking to give some additional nodes as targets, just incase if one of the target node is down at the time of starting recovery, we could simply use one of the additional target node. We donlt need this in first version, but that is one thought. So, I feel we donlt need to mix that missing indexes with target node indexes. 
Missing indexes are nothing but the missing container index positions.

Reply:
Sure, let's keep it as missing container indexes.
And another small question, any special reason to have `List<Long>` for `srcNodesIndexes`, and `byte[]` for `missingContainerIndexes`?

Reply:
srcNodes Indexes are the indexes where are nodes available. Which are nothing but replica indexes. We could have a map also for srcNode and srcNode Index, but looks like we need to have another custom proto class to represent that. Just avoiding another class for this purpose.

Reply:
Its a bit strange to have a List<Long> for srcIndexes and byte[] for missing? Would it be better to have a List<Int> for both to be consistent? I don't think they need to be long, as the index will never be more than 14 for 10-4.

I also worry about some bug resulting in a different size of srcNodesIndexes and srcDatanodes, or the order being wrong. What about this command receiving List<ContainerReplica> - that object contains both the index and datanodeDetails, so we avoid a map and just pass 1 parameter instread of two.

When forming the protobuf, we don't need to send ContainerReplica - we can still send replicaIndex and DatanodeDetails as they are now.

Reply:
I think this command can receive the ContainerReplica, but we don't need to send it all over the RPC. Just pull the index and datanodeDetails out of it and send it over. The command on the DN side receives the proto message I think, so its parameters can be different.

In replicationManager, where this command will be triggered, we will have the list of existing ContainerReplica, so they can be passed to this command without any transformation at the replicationManager side.

Reply:
Looks like we need move ContainerReplica to container-service module. Currently its part of scm-server only. 
Currently commands are part of the container-service module. 
I am not sure its worth doing that. It may be simple to create command with it's required info. Let me know if that's ok with you.

Reply:
OK - probably not worth the effort to start moving classes around to just save a parameter in the method spec.

Reply:
@umamaheswararao thanks for the work! 
i am not sure why `byte[]` for missing container index? seems int[] or something like this may be more easy to understand? or is there any special reason for `byte[]`? 
NIT:  i prefer `missingReplicaIndexes ` instead of `missingContainerIndexes` , since the specific replica of container is missing actually, not container itself. 

for the current implementation of ozone EC , we don`t shuffle the blocks among EC container group , it means some containers will only have blocks of data chunks , and others will only have blocks of parity chunks.  shuflle can scatter the read of data blocks to different datanode, so i believe it has some significance for the performance of the whole ozone cluster , and we need to support it in the future. at that time , the replica index will not be equal to the chunk index in a stripe , and we need some other mechanism to identify the chunk index in a stripe.

Reply:
> shuflle can scatter the read of data blocks to different datanode, so i believe it has some significance for the performance of the whole ozone cluster , and we need to support it in the future. 
There is no plan to try to support this. It makes things too complex for negligible gains. We only have one container per pipeline, so each time a new container is allocated, it gets new nodes in a random order. So over many containers we will see the data fairly evenly spread without having to worry about shuffling writes inside one container group.

Reply:
When you say containerIndex, it means one of the index position container missed. Not saying "MissingContainer" only. Example, we don't call ReplicateReplica, instead we are calling ReplicateContainer. Most of the times, Replica represents, the same copy. In EC each replica is not a same copy. However we tried to reuse parts like ConrainerReplica etc, so named it as ReplicaIndexes. That is the reason, I just put as missingContainerIndex, as Class name is saying "ReconstructECContainer". Let's me know if that is ok.